src/battle_util.c

```diff
@@ -6236,12 +6202,22 @@ u8 AbilityBattleEffects(u8 caseID, u8 battler, u16 ability, u8 special, u16 move
         }
         break;
     case ABILITYEFFECT_ON_TERRAIN:  // For ability effects that activate when the field terrain changes.
-        battler = gBattlerAbility = gBattlerAttacker = gBattleScripting.battler;
+        battler = gBattlerAbility = gBattleScripting.battler;
         switch (GetBattlerAbility(battler))
         {
+        case ABILITY_MIMICRY:
+            if (!gSpecialStatuses[battler].terrainAbilityDone)
```

Comment:
I suggest: `if (!gSpecialStatuses[battler].terrainAbilityDone && ChangeTypeBasedOnTerrain(battler))` so if there's no active terrain, mimicry won't activate

Reply:
Same as above. Will push, but it feels weird to have 2 calls to the function together, one for a check and the other one to execute an operation.